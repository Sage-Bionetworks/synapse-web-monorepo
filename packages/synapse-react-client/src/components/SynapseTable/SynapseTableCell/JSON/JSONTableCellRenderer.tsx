import React from 'react'
import { Typography } from '@mui/material'
import {
  isJSONObjectAllPrimitiveKeys,
  isJSONPrimitive,
} from './JSONRendererUtils'
import { JSONPrimitiveRenderer } from './JSONPrimitiveRenderer'
import { JSONArrayRenderer } from './JSONArrayRenderer'
import { JSONObjectRenderer } from './JSONObjectRenderer'
import { ComplexJSONRenderer } from './ComplexJSONRenderer'

export type JSONTableCellRendererProps = {
  value: string | null
}

export default function JSONTableCellRenderer(
  props: JSONTableCellRendererProps,
) {
  let value: unknown = props.value
  try {
    if (typeof value === 'string') {
      value = JSON.parse(value)
    }
  } catch (e) {
    console.warn('JSONTableCellRenderer: failed to parse JSON', e)
  }

  if (isJSONPrimitive(value) || value === null) {
    return (
      <Typography variant="smallText1">
        <JSONPrimitiveRenderer value={value} />
      </Typography>
    )
  } else if (Array.isArray(value) && value.every(isJSONPrimitive)) {
    return <JSONArrayRenderer value={value} />
  } else if (
    typeof value === 'object' &&
    // To render a simple flat table, ensure that we have no array/object values
    isJSONObjectAllPrimitiveKeys(value)
  ) {
    return <JSONObjectRenderer value={value} />
  }

  return <ComplexJSONRenderer value={value} />
}
