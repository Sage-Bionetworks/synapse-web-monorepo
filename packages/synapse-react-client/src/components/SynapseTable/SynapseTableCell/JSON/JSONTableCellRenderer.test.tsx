import JSONTableCellRenderer from './index'
import { render } from '@testing-library/react'

describe('JSONTableCellRenderer', () => {
  it.each<[string, any]>([
    ['Empty array', []],
    ['Empty object', {}],
    ['Array of primitives', ['foo', 1000, -25.67, null, true, false]],
    [
      'Object of primitives',
      {
        string: 'foo',
        numberInt: 1000,
        numberFloat: -25.67,
        null: null,
        true: true,
        false: false,
      },
    ],
    ['one item', ['foo']],
    ['one null item', [null]],
    ['null', null],
    [
      'complex array',
      [
        {
          foo: [
            'bar',
            'baz',
            {
              qux: 'quux',
              lorem: 'ipsum',
            },
          ],
        },
        null,
        ['qwerty', 'asdf'],
      ],
    ],
    [
      'complex object',
      {
        foo: [
          'bar',
          'baz',
          {
            qux: ['quux', 'abc'],
            lorem: 'ipsum',
          },
        ],

        null: null,
        bar: 'baz',
        zxcv: {
          querty: 'asdf',
        },
      },
    ],
  ])('handles %s case', (_case, jsonObject) => {
    const { container } = render(<JSONTableCellRenderer value={jsonObject} />)
    expect(container).toMatchSnapshot()
  })
})
