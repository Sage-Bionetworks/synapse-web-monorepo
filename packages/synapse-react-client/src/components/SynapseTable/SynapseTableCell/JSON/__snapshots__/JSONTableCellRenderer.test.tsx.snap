// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JSONTableCellRenderer > handles Array of primitives case 1`] = `
<div>
  <span
    class="MuiTypography-root MuiTypography-smallText1 css-mizb62-MuiTypography-root"
  >
    6 items
    <button
      class="ExpandableTableData__expandButton"
    >
      <span
        aria-label=""
        class="styled-svg-wrapper"
        data-mui-internal-clone-element="true"
        data-svg="addBoxOutline"
        id="icon-addBoxOutline"
        role="img"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-18y5paw-MuiSvgIcon-root"
          data-testid="AddBoxOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2m0 16H5V5h14zm-8-2h2v-4h4v-2h-4V7h-2v4H7v2h4z"
          />
        </svg>
      </span>
    </button>
  </span>
  <div
    class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-cwrbtg-MuiCollapse-root"
    style="min-height: 0px;"
  >
    <div
      class="MuiCollapse-wrapper MuiCollapse-vertical css-1x6hinx-MuiCollapse-wrapper"
    >
      <div
        class="MuiCollapse-wrapperInner MuiCollapse-vertical css-1i4ywhz-MuiCollapse-wrapperInner"
      >
        <div
          class="MuiBox-root css-70fdo6"
        >
          <span
            class="MuiTypography-root MuiTypography-smallText1 css-z2uw0f-MuiTypography-root"
          >
            foo
          </span>
        </div>
        <div
          class="MuiBox-root css-70fdo6"
        >
          <span
            class="MuiTypography-root MuiTypography-smallText1 css-z2uw0f-MuiTypography-root"
          >
            1,000
          </span>
        </div>
        <div
          class="MuiBox-root css-70fdo6"
        >
          <span
            class="MuiTypography-root MuiTypography-smallText1 css-z2uw0f-MuiTypography-root"
          >
            -25.67
          </span>
        </div>
        <div
          class="MuiBox-root css-70fdo6"
        >
          <span
            class="MuiTypography-root MuiTypography-smallText1 css-z2uw0f-MuiTypography-root"
          >
            <span
              class="MuiTypography-root MuiTypography-smallText1 SRC-inactive css-z2uw0f-MuiTypography-root"
            >
              &lt;null value&gt;
            </span>
          </span>
        </div>
        <div
          class="MuiBox-root css-70fdo6"
        >
          <span
            class="MuiTypography-root MuiTypography-smallText1 css-z2uw0f-MuiTypography-root"
          >
            true
          </span>
        </div>
        <div
          class="MuiBox-root css-70fdo6"
        >
          <span
            class="MuiTypography-root MuiTypography-smallText1 css-z2uw0f-MuiTypography-root"
          >
            false
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`JSONTableCellRenderer > handles Empty array case 1`] = `
<div>
  <span
    class="MuiTypography-root MuiTypography-smallText1 SRC-inactive css-z2uw0f-MuiTypography-root"
  >
    Empty array
  </span>
</div>
`;

exports[`JSONTableCellRenderer > handles Empty object case 1`] = `
<div>
  <span
    class="MuiTypography-root MuiTypography-smallText1 SRC-inactive css-z2uw0f-MuiTypography-root"
  >
    Empty object
  </span>
</div>
`;

exports[`JSONTableCellRenderer > handles Object of primitives case 1`] = `
<div>
  <div
    class="MuiBox-root css-34ji4n"
  >
    <span
      class="MuiTypography-root MuiTypography-smallText1 css-z2uw0f-MuiTypography-root"
    >
      string
    </span>
    <span
      class="MuiTypography-root MuiTypography-smallText1 css-ekwlr8-MuiTypography-root"
    >
      foo
    </span>
  </div>
  <div
    class="MuiBox-root css-34ji4n"
  >
    <span
      class="MuiTypography-root MuiTypography-smallText1 css-z2uw0f-MuiTypography-root"
    >
      numberInt
    </span>
    <span
      class="MuiTypography-root MuiTypography-smallText1 css-ekwlr8-MuiTypography-root"
    >
      1,000
    </span>
  </div>
  <div
    class="MuiBox-root css-34ji4n"
  >
    <span
      class="MuiTypography-root MuiTypography-smallText1 css-z2uw0f-MuiTypography-root"
    >
      numberFloat
    </span>
    <span
      class="MuiTypography-root MuiTypography-smallText1 css-ekwlr8-MuiTypography-root"
    >
      -25.67
    </span>
  </div>
  <div
    class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-cwrbtg-MuiCollapse-root"
    style="min-height: 0px;"
  >
    <div
      class="MuiCollapse-wrapper MuiCollapse-vertical css-1x6hinx-MuiCollapse-wrapper"
    >
      <div
        class="MuiCollapse-wrapperInner MuiCollapse-vertical css-1i4ywhz-MuiCollapse-wrapperInner"
      >
        <div
          class="MuiBox-root css-34ji4n"
        >
          <span
            class="MuiTypography-root MuiTypography-smallText1 css-z2uw0f-MuiTypography-root"
          >
            null
          </span>
          <span
            class="MuiTypography-root MuiTypography-smallText1 css-ekwlr8-MuiTypography-root"
          >
            <span
              class="MuiTypography-root MuiTypography-smallText1 SRC-inactive css-z2uw0f-MuiTypography-root"
            >
              &lt;null value&gt;
            </span>
          </span>
        </div>
        <div
          class="MuiBox-root css-34ji4n"
        >
          <span
            class="MuiTypography-root MuiTypography-smallText1 css-z2uw0f-MuiTypography-root"
          >
            true
          </span>
          <span
            class="MuiTypography-root MuiTypography-smallText1 css-ekwlr8-MuiTypography-root"
          >
            true
          </span>
        </div>
        <div
          class="MuiBox-root css-34ji4n"
        >
          <span
            class="MuiTypography-root MuiTypography-smallText1 css-z2uw0f-MuiTypography-root"
          >
            false
          </span>
          <span
            class="MuiTypography-root MuiTypography-smallText1 css-ekwlr8-MuiTypography-root"
          >
            false
          </span>
        </div>
      </div>
    </div>
  </div>
  <span
    class="MuiTypography-root MuiTypography-smallText1 css-a8zhlg-MuiTypography-root"
  >
    3 more
    <button
      class="ExpandableTableData__expandButton"
    >
      <span
        aria-label=""
        class="styled-svg-wrapper"
        data-mui-internal-clone-element="true"
        data-svg="addBoxOutline"
        id="icon-addBoxOutline"
        role="img"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-18y5paw-MuiSvgIcon-root"
          data-testid="AddBoxOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2m0 16H5V5h14zm-8-2h2v-4h4v-2h-4V7h-2v4H7v2h4z"
          />
        </svg>
      </span>
    </button>
  </span>
</div>
`;

exports[`JSONTableCellRenderer > handles complex array case 1`] = `
<div>
  <span
    class="MuiTypography-root MuiTypography-smallText1 css-rrx1sp-MuiTypography-root"
  >
    Expand
     all
    <button
      class="ExpandableTableData__expandButton"
    >
      <span
        aria-label=""
        class="styled-svg-wrapper"
        data-mui-internal-clone-element="true"
        data-svg="addBoxOutline"
        id="icon-addBoxOutline"
        role="img"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-18y5paw-MuiSvgIcon-root"
          data-testid="AddBoxOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2m0 16H5V5h14zm-8-2h2v-4h4v-2h-4V7h-2v4H7v2h4z"
          />
        </svg>
      </span>
    </button>
  </span>
  <div
    class="MuiBox-root css-yc55n3"
  >
    <ol
      role="tree"
      style="padding: 0px; margin: 0px; list-style-type: none;"
    >
      <li
        aria-expanded="false"
        role="treeitem"
        style="color: black; line-height: 1.2; cursor: default; box-sizing: border-box; list-style: none; font-family: Menlo, monospace; font-size: 11px;"
      >
        <div>
          <span
            style="color: rgb(110, 110, 110); display: inline-block; font-size: 12px; margin-right: 3px; user-select: none; transform: rotateZ(0deg);"
          >
            ▶
          </span>
          <span
            style="font-style: italic;"
          >
            (3) 
          </span>
          <span
            style="font-style: italic;"
          >
            [
            <span>
              Object
            </span>
            , 
            <span
              style="color: rgb(128, 128, 128);"
            >
              null
            </span>
            , 
            <span>
              Array(2)
            </span>
            ]
          </span>
        </div>
        <ol
          role="group"
          style="margin: 0px; padding-left: 12px;"
        />
      </li>
    </ol>
  </div>
</div>
`;

exports[`JSONTableCellRenderer > handles complex object case 1`] = `
<div>
  <span
    class="MuiTypography-root MuiTypography-smallText1 css-rrx1sp-MuiTypography-root"
  >
    Expand
     all
    <button
      class="ExpandableTableData__expandButton"
    >
      <span
        aria-label=""
        class="styled-svg-wrapper"
        data-mui-internal-clone-element="true"
        data-svg="addBoxOutline"
        id="icon-addBoxOutline"
        role="img"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-18y5paw-MuiSvgIcon-root"
          data-testid="AddBoxOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2m0 16H5V5h14zm-8-2h2v-4h4v-2h-4V7h-2v4H7v2h4z"
          />
        </svg>
      </span>
    </button>
  </span>
  <div
    class="MuiBox-root css-yc55n3"
  >
    <ol
      role="tree"
      style="padding: 0px; margin: 0px; list-style-type: none;"
    >
      <li
        aria-expanded="false"
        role="treeitem"
        style="color: black; line-height: 1.2; cursor: default; box-sizing: border-box; list-style: none; font-family: Menlo, monospace; font-size: 11px;"
      >
        <div>
          <span
            style="color: rgb(110, 110, 110); display: inline-block; font-size: 12px; margin-right: 3px; user-select: none; transform: rotateZ(0deg);"
          >
            ▶
          </span>
          <span
            style="font-style: italic;"
          />
          <span
            style="font-style: italic;"
          >
            {
            <span>
              <span
                style="color: rgb(136, 19, 145);"
              >
                foo
              </span>
              : 
              <span>
                Array(3)
              </span>
            </span>
            , 
            <span>
              <span
                style="color: rgb(136, 19, 145);"
              >
                null
              </span>
              : 
              <span
                style="color: rgb(128, 128, 128);"
              >
                null
              </span>
            </span>
            , 
            <span>
              <span
                style="color: rgb(136, 19, 145);"
              >
                bar
              </span>
              : 
              <span
                style="color: rgb(196, 26, 22);"
              >
                "
                baz
                "
              </span>
            </span>
            , 
            <span>
              <span
                style="color: rgb(136, 19, 145);"
              >
                zxcv
              </span>
              : 
              <span>
                Object
              </span>
            </span>
            }
          </span>
        </div>
        <ol
          role="group"
          style="margin: 0px; padding-left: 12px;"
        />
      </li>
    </ol>
  </div>
</div>
`;

exports[`JSONTableCellRenderer > handles null case 1`] = `
<div>
  <span
    class="MuiTypography-root MuiTypography-smallText1 css-z2uw0f-MuiTypography-root"
  >
    <span
      class="MuiTypography-root MuiTypography-smallText1 SRC-inactive css-z2uw0f-MuiTypography-root"
    >
      &lt;null value&gt;
    </span>
  </span>
</div>
`;

exports[`JSONTableCellRenderer > handles one item case 1`] = `
<div>
  <span
    class="MuiTypography-root MuiTypography-smallText1 css-z2uw0f-MuiTypography-root"
  >
    foo
  </span>
</div>
`;

exports[`JSONTableCellRenderer > handles one null item case 1`] = `
<div>
  <span
    class="MuiTypography-root MuiTypography-smallText1 css-z2uw0f-MuiTypography-root"
  >
    <span
      class="MuiTypography-root MuiTypography-smallText1 SRC-inactive css-z2uw0f-MuiTypography-root"
    >
      &lt;null value&gt;
    </span>
  </span>
</div>
`;
